core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -35,92 +35,109 @@ oppia.factory('NumberWithUnitsObjectFactory', [
       this.type = type;
       this.real = real;
       this.fraction = fractionObj;
-      this.units = unitsObj;
+      this.units = unitsObj.units;
     };
 
     NumberWithUnits.prototype.toString = function() {
       var numberWithUnitsString = '';
-      if (this.units.toString().includes('$')) {
-        this.units.units = this.units.toString().replace('$', '');
+      var unitsString = UnitsObjectFactory.fromList(this.units).toString();
+      if (unitsString.includes('$')) {
+        unitsString = unitsString.replace('$', '');
         numberWithUnitsString += '$' + ' ';
       }
-      if (this.units.toString().includes('Rs')) {
-        this.units.units = this.units.toString().replace('Rs', '');
+      if (unitsString.includes('Rs')) {
+        unitsString = unitsString.replace('Rs', '');
         numberWithUnitsString += 'Rs' + ' ';
       }
-      if (this.units.toString().includes('₹')) {
-        this.units.units = this.units.toString().replace('₹', '');
+      if (unitsString.includes('₹')) {
+        unitsString = unitsString.replace('₹', '');
         numberWithUnitsString += '₹' + ' ';
       }
 
       if (this.type === 'real') {
         numberWithUnitsString += this.real + ' ';
-      } else {
+      } else if (this.type === 'fraction') {
         numberWithUnitsString += this.fraction.toString() + ' ';
       }
-      numberWithUnitsString += this.units.toString();
+      numberWithUnitsString += unitsString.trim();
       numberWithUnitsString = numberWithUnitsString.trim();
 
       return numberWithUnitsString;
     };
 
+    NumberWithUnits.prototype.toDict = function() {
+      return {
+        type: this.type,
+        real: this.real,
+        fraction: this.fraction.toDict(),
+        units: this.units.toDict().units
+      };
+    };
+
     NumberWithUnits.fromRawInputString = function(rawInput) {
       rawInput = rawInput.trim();
       var type = '';
-      var real = '';
-      var fractionObj = '';
+      var real = 0.0;
+      // Default fraction value.
+      var fractionObj = FractionObjectFactory.fromRawInputString('0/1');
       var units = '';
-      var value = 0;
-      var unitObj = '';
+      var value = '';
+      var unitObj = [];
```

Comment:
nit: default value for object should be `null`

Reply:
As  we concluded at the meeting that it should always return `NumberWithUnits` objects so, I took this default value. We can not just return `null` as it will not match the return type.

Reply:
In that case, check for the null value possibility in initialization function (and if unitsObj is null then set units to []) because calling `unitsObj.units` when unitsObj is [] will generate errors.

Reply:
See line 141. This default unitObj is first passed to `UnitsObjectFactory` to get corresponding units object and then, passed to `NumberWithUnits` constructor. So, this behaviour seems correct to me.

Reply:
Oh, I see. Then can we not assign the value till line 141. It seems that this initialization is redundant?

Reply:
Yes, we can. But generally isn't it a good idea to initialize every variable?

Reply:
You can do it either way but it is generally a good idea to initialize variables only at the point when you need them.